src/pull_request.rs

```diff
@@ -0,0 +1,368 @@
+use crate::db::*;
+use crate::{
+	error,
+	github,
+	github_bot::GithubBot,
+	matrix,
+	matrix_bot::MatrixBot,
+	project,
+	Result,
+};
+use rocksdb::DB;
+use snafu::{
+	GenerateBacktrace,
+	ResultExt,
+};
+use std::collections::HashMap;
+use std::time::{
+	Duration,
+	SystemTime,
+};
+
+/*
+ * Ping periods measured in seconds
+ */
+const STATUS_FAILURE_PING_PERIOD: u64 = 3600 * 24;
+const ISSUE_NOT_ASSIGNED_PING_PERIOD: u64 = 3600 * 24;
+
+const FALLBACK_ROOM_ID: &'static str = "!aenJixaHcSKbJOWxYk:matrix.parity.io";
+const ISSUE_MUST_EXIST_MESSAGE: &'static str =
+	"Every pull request must address an issue.";
+const ISSUE_ASSIGNEE_NOTIFICATION: &'static str = "{1} addressing {2} has been opened by {3}. Please reassign the issue or close the pull request.";
+const REQUESTING_REVIEWS_MESSAGE: &'static str = "{1} is in need of reviewers.";
+const STATUS_FAILURE_NOTIFICATION: &'static str =
+	"{1} has failed status checks.";
+
+fn require_reviewer(
+	pull_request: &github::PullRequest,
+	repo: &github::Repository,
+	matrix_bot: &MatrixBot,
+	github_to_matrix: &HashMap<String, String>,
+	project_info: Option<&project::ProjectInfo>,
+) {
+	let author_is_owner = project_info
+		.and_then(|p| p.owner.as_ref())
+		.map(|u| u == &pull_request.user.login)
+		.unwrap_or(false);
+	let author_is_delegated = project_info
+		.and_then(|p| p.delegated_reviewer.as_ref())
+		.map(|u| u == &pull_request.user.login)
+		.unwrap_or(false);
+
+	if !author_is_delegated {
+		// TODO
+		// require review from delegated reviewer
+	} else if !author_is_owner {
+		// TODO
+		// require review from project owner
+	} else {
+		// post a message in the project's Riot channel, requesting a review;
+		// repeat this message every 24 hours until a reviewer is assigned.
+		if let Some(ref room_id) =
+			&project_info.and_then(|p| p.matrix_room_id.as_ref())
+		{
+			matrix_bot.send_public_message(
+				&room_id,
+				&REQUESTING_REVIEWS_MESSAGE
+					.replace("{1}", &format!("{}", pull_request.html_url)),
+			);
+		} else {
+			matrix_bot.send_public_message(
+				&FALLBACK_ROOM_ID,
+				&REQUESTING_REVIEWS_MESSAGE
+					.replace("{1}", &format!("{}", pull_request.html_url)),
+			);
+		}
+	}
+
+	/*
+	 * if they are not the Delegated Reviewer (by default the project owner),
+	 * then Require a Review from the Delegated Reviewer; otherwise, if the
+	 * author is not the project owner, then Require a Review from the
+	 * Project Owner; otherwise, post a message in the project's Riot
+	 * channel, requesting a review; repeat this message every 24 hours until
+	 * a reviewer is assigned.
+	 */
```

Comment:
It seems this comment is supposed to be on `fn require_reviewer`?

Reply:
Yeah it's copied from the spec, I'll remove it when the function body is complete.